### ISSUE TITLE
Macpass adding folders (duplicating itself)

### ISSUE THREAD

**Issue by oehm-smith** [MacPass/MacPass#818]
**Describe the bug**
I've been a MP user for many years.  I have one main database I've been using and over time the folder structure is reproduced in a sub-folder.  Over and over.  See the attached picture.  Within each of the 'database' folders is an exact clone of the folder structure.  The actual entries are spread out in the different folders as I (and my wife) have at times not realised this was happening, or we had to creaate an entry somewhere.

<img width="253" alt="2018-07-27_macpass_issue" src="https://user-images.githubusercontent.com/2556595/43292936-156f84aa-917b-11e8-8eb3-93f8f6d57408.png">

**To Reproduce**
I don't know what causes this. I suspect its when the same file used on multiple Macs.

**Expected behavior**
The folder structure should remain fixed at what the user creates.

**Screenshots**
See description above.

**Desktop (please complete the following information):**
 - macOS 10.13.6 (17G65) - though this has been happening for a long time
 - MP Version 0.7.3 (17830)

**Additional context**
The problem may be difficult to work out or something isolated to a few users.  A good solution for me would be a script to clean this up.

**Comment by mstarke** [comment]
This might be a bug in KDB serialization. Do you use the KDBX or KDB file format?

**Comment by oehm-smith** [comment]
KDB - This is the original databases I used with KeepassX.  I can see benefit in converting to kdbx though is there a way then to merge any entries scattered across the various folders and kill the duplicates?

**Comment by mstarke** [comment]
This seems to be an issue with merging changed databases. KDB does not hold a Root folder so it is generated each time the Database is loaded. This seems to be the culprit for the duplication, although that should not result ins duplication of entries. I'll take a look and see if I can fix this ASAP. Also I'll try to device a way for you to remove all those duplicates without loosing any data.

**Comment by oehm-smith** [comment]
Great thanks Michael.

**Comment by mstarke** [comment]
I'm onto adding the feature to merge KDB databases on external change but this is not ideal. I'm pondering of disabling this altogether since groups can only be identified via their properties. This leads to the situations a group might map to different groups and it's not possible (at least with adequate complexity) to determine a unique mapping.

You should upgrade your files to KDBX and everything should work out. As said before, there's no way to automatically remove the duplicates so you'll have to manually sort through the groups to manage those. I'm sorry for the fuck up but i forgot to correctly work with KDB files on merge.

I'll update this issue on the progress. Additionally, there's a related on on KeePassKit (https://github.com/MacPass/KeePassKit/issues/39)

**Comment by mstarke** [comment]
@oehm-smith could you try to run the setup with the current continuous build? This version tries to map folders based on names. This will still create problems, e.g. if you have duplicate names. But no duplicates will be created it the name is already present. Be aware that the match is **only** name based. And again a big sorry for messing up you folder structure

**Comment by oehm-smith** [comment]
(I replied this in email yesterday but it didn't seem to make it here.  So adding directly)

Hi Michael,

What will I gain from running the latest version?  What will happen with duplicates given the 'problems'?  Will I still need to manually resolve the duplicate groups (eg. consolidate entries) as you suggested in a previous post?  And if so might it be just as good to do this fix and update to next version when its a proper release?  Or won't this problem (in KDBX - I saved as this already) be fixed until that new code is used?  And so any manual work will need to be redone at a later date (when the release comes out)?  Apologies for so many questions.

Thanks,

Brooke

**Comment by mstarke** [comment]
Brooke,

The latest version will not clean up any duplicates. It will however prevent any duplicates from showing up when you merge KDB files. You will have to fix the problem manually but it should not happen anymore with KDB files. Since you saved the file as KDBX you are now save and no duplication will happen again on merge.

The reason for asking this is to have more than my test setup to see if it actually works. If you feel comfortable of trying this setup with your previous KDB setup it would be awesome. Since I have found the culprit and the issue should not occur again but since the fix is not perfect (as described, it cannot be) there might be some corner cases that aren't handled very well.

TL;DR; If you have switched to KDBX there's no need to use a newer version. Just stick with 0.7.3 until 0.7.4 (or even 0.8) get's released.

Again, I'm very sorry for the inconvenience my error has created.

**Comment by mstarke** [comment]
The only thing that should not happen is more duplicates showing up. Also, created entries on the other end should end up in the same location on both databases.

Thank you very much for the effort!

**Comment by mstarke** [comment]
Works-for-all solution by matching groups by name is working. Refer to [KeePassKit #39]( https://github.com/MacPass/KeePassKit/issues/39) for details.

**Comment by oehm-smith** [comment]
Thanks a lot Michael,

You're a champ!

Cheers,

Brooke

**Comment by oehm-smith** [comment]
Hi Michael,

I finally found time to clean up my database.  I tested it and it works as well.  I can now see the sync / merge functionality working.  So nice.

Cheers,

Brooke

**Comment by mstarke** [comment]
Thank you for testing it out again. I've since started adding custom extensions to back-port features of KDBX to KDB but it's not final. This will however only allow for MacPass to better merge KDB files. All other clients will probably never support the custom extensions.

**Comment by oehm-smith** [comment]
Sounds good.  KDB is legacy right?  Perhaps the problem is that Linux 
doesn't support KDBX.

**Comment by mstarke** [comment]
AFAIK the most used Linux client is KeePassXC which does support KDBX and allows to import KBD so this shouldn't be an issue. There are some Android clients that only support KDB but all "relevant" platforms have at least one decent client that supports KDBX.

**Comment by mstarke** [comment]
I'm sorry but I have no clue what you're trying to say in your first sentence. Would you be so kind to explain it to me? But it's vital to keep KDB use cases in mind since your experience shows that if I do not cover all the use-cases strange things happen. And strange things happening to password containers is something I rather not do too often :). Of course pouring too much time into corner cases is a somewhat waste of time but most of those changes do not take too much time and enhance the user experience dramatically so I'm tempted to implement them even though it'll only show up in MacPass.

**Comment by oehm-smith** [comment]
I meant its nice that you are considering all options (the corner cases).  I just wondered whether this work is really required - in your reply you said that you feel its important to enhance the user experience.  All good.  Thanks for your efforts.
